### PR TITLE
Iterator

### PR DESCRIPTION
**Pull request by magnus-madsen**

### CODE REVIEW

**Comment by stephentetley** [comment]
Hi Magnus, with the `Iterator` module, have you considered a `Close` operation to be run when the iterator is exhausted as well as `Next` and `Done`?

In most circumstances `Close` would do nothing, but for e.g. JDBC database connections / resultsets it would call the respective close methods in Java-land. I think some types of file handle and other objects would also need closing.

I've a similar library here that includes `Close`:

https://github.com/stephentetley/flix-sandbox/blob/master/src/flix/toplevel/Cursor.flix

I haven't really worked out the lifecycle of a Cursor - i.e all functions that exhaust the Cursor should close it, but I didn't account for e.g. just taking a prefix with `take` which would leave the cursor open.

So I could implement `filter` I used version of the `Step` datatype from the Haskell Stream fusion paper which might be relevant:

http://citeseer.ist.psu.edu/viewdoc/summary?doi=10.1.1.104.7401

**Comment by magnus-madsen** [comment]
> Hi Magnus, with the `Iterator` module, have you considered a `Close` operation to be run when the iterator is exhausted as well as `Next` and `Done`?

No, I never thought about that. Thanks for bringing it up!

I guess the last call to `next()` could in principle release the underlying resource (could be a JDBC connection but equally well a file handle). But indeed, as you point out, what happens when you don't consume the entire iterator?

For some computations you may not even want to consume the entire iterator but you may still want to release its resources!

I will need to think a bit. I am currently learning a lot about streams and lazy evaluation for some research, but I did not consider this specific problem.

Perhaps the resource management should really be handled separately. When we have some notion of regions (see other PR- which is also just an experiment) then we could have the region system automatically close an iterator (backed by e.g. a file) as soon as it goes out of scope. Then its more a matter of having a `Closeable` type class and having the compiler inserts calls to `close` in the right places, and having `Iterator` (or some iterators?) implement `Closeable`.

**Comment by magnus-madsen** [comment]
I also should briefly mention there are both PullStreams, PushStreams, LazyLists, Iterators, Iteratees, and Cursors (?) -- its quiet a bit to wrap your head around while trying to do something useful and new! :)

**Comment by stephentetley** [comment]
I found Iteratees and the first sprint of activity in the Haskell world (pipes and conduits) very hard to follow - especially as at the time I didn't need to use streams for anything. 

I thought sections 1 to 3 of this paper were very good at clarifying the state of subject:

https://yanniss.github.io/streams-popl17.pdf

I think I'm right in saying both your Iterator and my Cursor (plus my PullStream [1]) are Pull Streams as they are essentially objects with two methods `init` and `next` (like the List function `unfold`). Pull Streams are more expressive than Push Streams as you can't define `zip` on a Push Stream (page 4 of the paper). I actually called my Cursor module "Cursor" because I didn't want to call it "Iterator". 

[1] https://github.com/stephentetley/flix-sandbox/blob/master/src/flix/toplevel/PullStream.flix

**Comment by magnus-madsen** [comment]
(Our discussion can continue in some new tickets).
